### DISCUSSION TITLE
With Sinatra 4/Rack3, now requires "request.body.rewind"

### DISCUSSION THREAD

**Discussion by matobinder** [sinatra/sinatra#1999] [Q&A] [answered]
I'm always a bit fuzzy on "is this Sinatra or Rack?" I think this is Rack. but  this issue came about when I upgraded to Sinatra 4 (and get Rack 3 too)

So I'm running some dead simple rest services, that get JSON posts. I always just ran "request.body.read" to get the body. I've found some clients were NOT setting the content-type, so I've added a "request.body.rewind" before I do the read.

I'm not reading the body in my code, so guessing something else is doing this when there isn't a content-type.

So, as I don't want to try to make all my client "good players" by making them put content-type in.... is it a good rule of thumb to always do a "request.body.rewind"?

**Comment by dentarg** [comment, marked as answer]
Yeah, to be sure you get the body, you need to rewind first.

Note that "rack.input is no longer required to be rewindable." with Rack 3

https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md#rack-2--rack-3-compatibility

**Reply by matobinder** [reply]
Oops, I should have posted this in discussions. I just wasn't sure if something changed that was a bug, like I wasn't sure why I have to do a rewind now... but good to know I should have been doing that all along.
I'll dig around a bit more to see if I can figure out why/what is doing a request.body.read when there is no content type. I know its not in my code, but may some gem I am including is doing that

**Reply by matobinder** [reply]
Yay, I found where the rewind was removed, where it used to do it. All making more sense now.
https://github.com/rack/rack/commit/42aff22f708123839ba706cbe659d108b47c40c7
in lib/rack/request.rb, it used to do a rewind, now it doesn't, and that was for the change you mentioned

"rack.input is no longer required to be rewindable"
That really didn't make sense to me until I see why/what was done.

**Reply by dentarg** [reply]
The rack upgrade guide was just updated to include more info on this: https://github.com/rack/rack/pull/2153/files

**Reply by matobinder** [reply]
Thanks for updating the rack upgrade guide.

**Reply by dentarg** [reply]
That wasn't me, it was @dblock, I just happened to see it :)
